### PR TITLE
Working Network A, set weights issue for Network B-D

### PR DESCRIPTION
**Pull request by VMS-6511**

### CODE REVIEW

**Comment by yanndupis** [comment]
Hey @VMS-6511 , I have opened the pr #600. Once merged, it will fix the error you were getting with set_weights. With this small pr, I was able to run network b and the accuracy was close to 100%. Thank you.

File: examples/securenn/network_b.py [hunk 1]
```diff
@@ -4,23 +4,23 @@
...
 import sys
 from typing import List, Tuple
 
 import tensorflow as tf
+import tensorflow.keras as keras
 import tf_encrypted as tfe
 
 from examples.mnist.convert import get_data_from_tfrecord
 
 # tfe.set_tfe_events_flag(True)
```
Comment (yanndupis) [inline examples/securenn/network_b.py]: We can remove this commented line.

File: examples/securenn/network_b.py [hunk 2]
```diff
@@ -4,23 +4,23 @@
...
 
 if len(sys.argv) >= 2:
-  # config file was specified
+                                # config file was specified
```
Comment (yanndupis) [inline examples/securenn/network_b.py]: There are some extra spaces we should remove.

File: examples/securenn/network_b.py [hunk 3]
```diff
@@ -4,23 +4,23 @@
...
   config_file = sys.argv[1]
   config = tfe.config.load(config_file)
 else:
-  # default to using local config
+                                # default to using local config
```
Comment (yanndupis) [inline examples/securenn/network_b.py]: There are some extra spaces we should remove.

File: examples/securenn/network_b.py [hunk 4]
```diff
@@ -176,11 +156,11 @@ class PredictionClient():
...
   Args:
-    player_name: `str`, name of the `tfe.player.Player`
-                 representing the data owner
-    build_update_step: `Callable`, the function used to construct
-                       a local federated learning update.
-  """
+      player_name: `str`, name of the `tfe.player.Player`
+                  representing the data owner
+      build_update_step: `Callable`, the function used to
+                        construct a local federated learning update.
```
Comment (yanndupis) [inline examples/securenn/network_b.py]: There is an extra 2-space in this bloc.

File: examples/securenn/network_b.py [hunk 5]
```diff
@@ -127,18 +114,10 @@ def loop_body(i: tf.Tensor,
...
-      layer1 = pooling(tf.nn.relu(conv2d(x, Wconv1, self.STRIDE) + bconv1))
-      layer2 = pooling(tf.nn.relu(
-          conv2d(layer1, Wconv2, self.STRIDE) + bconv2))
-      layer2 = tf.reshape(layer2, [-1, self.HIDDEN_FC1])
-      layer3 = tf.nn.relu(tf.matmul(layer2, Wfc1) + bfc1)
-      logits = tf.matmul(layer3, Wfc2) + bfc2
-
-      loss = tf.reduce_mean(
-          tf.losses.sparse_softmax_cross_entropy(logits=logits, labels=y))
+      x = tf.reshape(x, [-1, 28, 28, 1])
```
Comment (yanndupis) [inline examples/securenn/network_b.py]: Instead of manually specifying 28, you could do `self.IN_DIM`

File: examples/securenn/network_b.py [hunk 6]
```diff
@@ -99,25 +74,37 @@ def build_training_graph(self, training_data) -> List[tf.Tensor]:
...
+    model.add(keras.layers.Conv2D(self.HIDDEN_CHANNELS,
+                                  (self.KERNEL, self.KERNEL),
+                                  batch_input_shape=(self.BATCH_SIZE,
+                                                     self.IN_DIM,
+                                                     self.IN_DIM,
+                                                     self.IN_CHANNELS)))
+    model.add(keras.layers.Activation('relu'))
+    model.add(keras.layers.AveragePooling2D())
+    model.add(keras.layers.Conv2D(
+        self.HIDDEN_CHANNELS, (self.KERNEL, self.KERNEL)))
```
Comment (yanndupis) [inline examples/securenn/network_b.py]: Let's keep the indentation consistent for the `Conv2d`. You can use the same approach as Con2d above.

File: examples/securenn/network_b.py [hunk 7]
```diff
@@ -193,8 +173,9 @@ def provide_input(self) -> List[tf.Tensor]:
 
     with tf.name_scope('pre-processing'):
       prediction_input = tf.reshape(
-          prediction_input, shape=(self.BATCH_SIZE, 784))
-      expected_result = tf.reshape(expected_result, shape=(self.BATCH_SIZE,))
+          prediction_input, shape=(self.BATCH_SIZE, 28, 28, 1))
```
Comment (yanndupis) [inline examples/securenn/network_b.py]: You could also use `self.IN_DIM` here.

File: examples/securenn/network_c.py [hunk 8]
```diff
@@ -101,25 +76,42 @@ def build_training_graph(self, training_data) -> List[tf.Tensor]:
...
-                              self.HIDDEN_C2], 1.)
-    bconv2 = bias_variable([1, 1, self.HIDDEN_C2])
-    wfc1 = weight_variable([self.HIDDEN_FC1, self.HIDDEN_FC2], 1.)
-    bfc1 = bias_variable([self.HIDDEN_FC2])
-    wfc2 = weight_variable([self.HIDDEN_FC2, self.OUT_N], 1.)
-    bfc2 = bias_variable([self.OUT_N])
-    params = [wconv1, bconv1, wconv2, bconv2, wfc1, bfc1, wfc2, bfc2]
+    model = keras.Sequential()
+    model.add(keras.layers.Conv2D(
+        self.HIDDEN_C1,
```
Comment (yanndupis) [inline examples/securenn/network_c.py]: Let's keep the indentation similar to network b unless for some reasons, it doesn't work.

File: examples/securenn/network_c.py [hunk 9]
```diff
@@ -101,25 +76,42 @@ def build_training_graph(self, training_data) -> List[tf.Tensor]:
...
+         self.KERNEL),
+        batch_input_shape=(
+            self.BATCH_SIZE,
+            self.IN_DIM,
+            self.IN_DIM,
+            self.IN_CHANNELS)))
+    model.add(keras.layers.Activation('relu'))
+    model.add(keras.layers.AveragePooling2D())
+    model.add(keras.layers.Conv2D(
+        self.HIDDEN_C2, (self.KERNEL, self.KERNEL)))
```
Comment (yanndupis) [inline examples/securenn/network_c.py]: Conv2d indentation.

File: examples/securenn/network_c.py [hunk 10]
```diff
@@ -128,18 +120,10 @@ def loop_body(i: tf.Tensor,
...
-      layer1 = pooling(tf.nn.relu(conv2d(x, Wconv1, self.STRIDE) + bconv1))
-      layer2 = pooling(tf.nn.relu(
-          conv2d(layer1, Wconv2, self.STRIDE) + bconv2))
-      layer2 = tf.reshape(layer2, [-1, self.HIDDEN_FC1])
-      layer3 = tf.nn.relu(tf.matmul(layer2, Wfc1) + bfc1)
-      logits = tf.matmul(layer3, Wfc2) + bfc2
-
-      loss = tf.reduce_mean(
-          tf.losses.sparse_softmax_cross_entropy(logits=logits, labels=y))
+      x = tf.reshape(x, [-1, 28, 28, 1])
```
Comment (yanndupis) [inline examples/securenn/network_c.py]: `self.IN_DIM` instead of 28.

File: examples/securenn/network_c.py [hunk 11]
```diff
@@ -193,8 +178,9 @@ def provide_input(self) -> List[tf.Tensor]:
 
     with tf.name_scope('pre-processing'):
       prediction_input = tf.reshape(
-          prediction_input, shape=(self.BATCH_SIZE, 784))
-      expected_result = tf.reshape(expected_result, shape=(self.BATCH_SIZE,))
+          prediction_input, shape=(self.BATCH_SIZE, 28, 28, 1))
```
Comment (yanndupis) [inline examples/securenn/network_c.py]: `self.IN_DIM`

File: examples/securenn/network_d.py [hunk 12]
```diff
@@ -100,38 +75,47 @@ def build_training_graph(self, training_data) -> List[tf.Tensor]:
...
-      # model construction
-      x = tf.reshape(x, [-1, self.IN_DIM, self.IN_DIM, 1])
-      layer1 = pooling(tf.nn.relu(conv2d(x, Wconv1, self.STRIDE) + bconv1))
-      layer1 = tf.reshape(layer1, [-1, self.HIDDEN_FC1])
-      layer2 = tf.nn.relu(tf.matmul(layer1, Wfc1) + bfc1)
-      logits = tf.matmul(layer2, Wfc2) + bfc2
-
-      loss = tf.reduce_mean(
-          tf.losses.sparse_softmax_cross_entropy(logits=logits, labels=y))
+      x = tf.reshape(x, [-1, 28, 28, 1])
```
Comment (yanndupis) [inline examples/securenn/network_d.py]: `self.IN_DIM`

File: examples/securenn/network_d.py [hunk 13]
```diff
@@ -185,8 +169,9 @@ def provide_input(self) -> List[tf.Tensor]:
 
     with tf.name_scope('pre-processing'):
       prediction_input = tf.reshape(
-          prediction_input, shape=(self.BATCH_SIZE, 1, 28, 28))
-      expected_result = tf.reshape(expected_result, shape=(self.BATCH_SIZE,))
+          prediction_input, shape=(self.BATCH_SIZE, 28, 28, 1))
```
Comment (yanndupis) [inline examples/securenn/network_d.py]: `self.IN_DIM`

**Comment by mortendahl** [comment]
Some of this stuff seems to overlap with https://github.com/tf-encrypted/tf-encrypted/pull/606 -- how do we avoid too many conflicts?

File: examples/securenn/network_c.py (hunk 8 above)
Comment (VMS-6511) [reply examples/securenn/network_c.py]: In the inference section this doesn't tend to work because I usually call the dimension attributes from `ModelTrainer`

**Comment by yanndupis** [comment]
@mortendahl, I think @VMS-6511 has just updated the the scripts, `network_a.py`, b, c and d. The other files is probably just after after a `git merge master` on this branch isn't it @VMS-6511 ? The other files seems to be in sync with master.

**Comment by VMS-6511** [comment]
Yeah I only updated the network files everything else should be as it is on master.
